server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -329,4 +290,101 @@ public void setLastAcceptedState(ClusterState clusterState) {
+     * between the IndicesStore deleting the index folder and the CoordinationState concurrently trying to write the updated
+     * metadata into it. We could probably solve this with careful synchronization, but in fact there is no need.  The persisted
+     * state on master-ineligible data nodes is mostly ignored - it's only there to support dangling index imports, which is
+     * inherently unsafe anyway. Thus we can safely delay metadata writes on master-ineligible data nodes until applying the
+     * cluster state, which is what DataOnlyNodePersistedState does:
+     */
+    static class DataOnlyNodePersistedState extends InMemoryPersistedState {
+
+        static final String APPLIER_UPDATE_THREAD_NAME = "DataOnlyNodePersistedState#updateTask";
+
+        private final IncrementalClusterStateWriter incrementalClusterStateWriter;
+        private final PrioritizedEsThreadPoolExecutor threadPoolExecutor;
+
+        private volatile ClusterState lastCommittedStateWithPersistence;
+
+        DataOnlyNodePersistedState(Settings settings, IncrementalClusterStateWriter incrementalClusterStateWriter,
+                                          ThreadPool threadPool) {
+            super(incrementalClusterStateWriter.getPreviousManifest().getCurrentTerm(),
+                incrementalClusterStateWriter.getPreviousClusterState());
+            this.incrementalClusterStateWriter = incrementalClusterStateWriter;
+            String nodeName = Objects.requireNonNull(Node.NODE_NAME_SETTING.get(settings));
+            this.threadPoolExecutor = EsExecutors.newSinglePrioritizing(
+                nodeName + "/" + APPLIER_UPDATE_THREAD_NAME,
+                daemonThreadFactory(nodeName, APPLIER_UPDATE_THREAD_NAME),
+                threadPool.getThreadContext(),
+                threadPool.scheduler());
+        }
+
+        @Override
+        public void markLastAcceptedStateAsCommitted() {
+            super.markLastAcceptedStateAsCommitted();
+            final ClusterState lastCommittedState = getLastAcceptedState();
+            if (lastCommittedState.blocks().disableStatePersistence()) {
+                return;
+            }
+
+            this.lastCommittedStateWithPersistence = lastCommittedState;
+
+            try {
+                // synchronously dereference any indices that are no longer referenced (to make sure that cleanup actions in
+                // IndicesClusterStateService / IndicesStore triggered by the exposed cluster state do not result in entries
+                // in the manifest to become orphaned).
+                incrementalClusterStateWriter.retainIndicesOnDataOnlyNode(lastCommittedState);
+            } catch (WriteStateException e) {
+                logger.error(new ParameterizedMessage("Failed to remove unreferenced indices for state with version {}",
+                    lastCommittedState.version()), e);
+                e.rethrowAsErrorOrUncheckedException();
```

Comment:
yes, I would like to be stricter in this specific case here, because if we just carry on, there is a risk for IndicesClusterStateService/IndicesStore to corrupt the data node's on-disk cluster state as it will remove folders of deleted indices while the manifest file still points at those.